WIP: fix(externalmetrics): removing strings.ToLower to be compatible to docs

sebbrandt87 opened the pull request:
### What does this PR do?

With this we remove the strings.ToLower(info.Metric) as we need to stay compatible to the DataDog [documentation] for custom metrics.

### Motivation

Without this fix the datadog agent cannot find a custom metric name like `customMetricName`, because it is looking for `custommetricname`.

### Additional Notes

Solves #10330

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

[documentation]: https://docs.datadoghq.com/metrics/custom_metrics/#naming-custom-metrics

sebbrandt87 commented:
seems I defined the test wrong first, that is why I repushed.

alai97 commented on releasenotes/notes/fix-externalmetrics-removing-strings-ToLower-68ce8a437cd681b0.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+fixes:
+  - |
+    Remove the strings.ToLower(info.Metric) to stay compatible to the
+    DataDog documentation for custom metrics.
```
```suggestion
    Removes `strings.ToLower(info.Metric)` to align with the custom metrics documentation.
```

alai97 reviewed (approved):
Looks good pending one copy nit!

sebbrandt87 commented:
I seem to not have set the `storeContent` in the unit test correct, might someone can give me a hint?

sebbrandt87 committed:
fix(externalmetrics): removing strings.ToLower to be compatible to docs

With this we remove the `strings.ToLower(info.Metric)` as we need to stay compatible to the DataDog [documentation] for custom metrics.

Without this fix the datadog agent cannot find a custom metric name like `customMetricName`, because it is looking for `custommetricname`.

Solves DataDog#10330

[documentation]: https://docs.datadoghq.com/metrics/custom_metrics/#naming-custom-metrics

vboulineau commented on pkg/clusteragent/externalmetrics/provider_test.go:
```diff
@@ -194,6 +195,31 @@ func TestGetExternalMetrics(t *testing.T) {
 			expectedError:           fmt.Errorf("DatadogMetric not found for metric name: nginx.net.request_per_s, datadogmetricid: default/dcaautogen-32402d8dfc05cf540928a606d78ed68c0607f7"),
 		},
+		{
+			desc: "Test CustomExternalMetric - DatadogMetric exists and is valid",
```
You may want to change the desc to something like "case sensitive metric name" or something

sebbrandt87 replied:
I will do that, thanks for the hint

sebbrandt87 replied:
Done

vboulineau commented:
> I seem to not have set the `storeContent` in the unit test correct, might someone can give me a hint? https://github.com/DataDog/datadog-agent/pull/10494/files#diff-dfb143c4813a794249396fe973647d30bd0b783d6977f5e222aadb2a4316afc9R200-R212

Seems to be passing now

sebbrandt87 commented:
then I found the correct unit test values 😉 
so I removed the WIP now 😄

sebbrandt87 commented:
Did a rebase against current main branch and changed unit test description to match the case content https://github.com/DataDog/datadog-agent/pull/10494#pullrequestreview-850195009

mx-psi commented:
I triggered Gitlab tests (running on my branch `mx-psi/pr-10494`). This will be merged once the current merge freeze is over

vboulineau commented on pkg/clusteragent/externalmetrics/provider.go:
```diff
@@ -113,9 +113,6 @@ func (p *datadogMetricProvider) GetExternalMetric(namespace string, metricSelect
 	log.Debugf("Received external metric query with ns: %s, selector: %s, metricName: %s", namespace, metricSelector.String(), info.Metric)
 
-	// Convert metric name to lower case to allow proper matching (and DD metrics are always lower case)
-	info.Metric = strings.ToLower(info.Metric)
```
Hey @sebbrandt87,

I further checked the current behaviour and in fact this won't solve your problem.

The `strings.ToLower` ensures we get lower case metric name, but in fact the HPA controller already sends the metricName as lowercase (so case is lost before reaching us), so it's more to replicate this behaviour in unit tests, etc.
You can verify that by activating DEBUG logs and checking the log line above.

What happens when you put an uppercase in your HPA's metricName is that an `autogen-` DatadogMetric is created keeping the case properly and the query made to Datadog keeps the case but this `autogen-` metric cannot be matched with incoming query due to this, leading a "not found" response from the Cluster Agent.

However, using a [custom DatadogMetric object](https://docs.datadoghq.com/agent/cluster_agent/external_metrics/?tab=helm#example-datadogmetric-object) allows to bypass this limitation and should fully support case correctly.

sebbrandt87 replied:
> What happens when you put an uppercase in your HPA's metricName is that an autogen- DatadogMetric is created keeping the case properly and the query made to Datadog keeps the case but this autogen- metric cannot be matched with incoming query due to this, leading a "not found" response from the Cluster Agent.

And this is the reason this PR was created and the unit test is proving that it is working with this change. Isn't it?

vboulineau replied:
Yes, but it's not relevant since the function will always get a lower case input, regardless of what you put inside the HPA.

sebbrandt87 replied:
This is a big migration issue for us. We need to enable CustomMetrics CRD and all our current HPA metrics pointed to a case sensitive custom metric break. As stated, custom metrics are allowed to be uppercase and so this fix here would match that behaviour.

vboulineau replied:
Hey @sebbrandt87,

I'll try to explain differently. Even we merge this change, it still won't work, because what is received by this function is always lowercase.
You may check that for yourself by activating `DEBUG` logs in the Cluster Agent and check for this log (line 114, so before `strings.ToLower` is called):
```
log.Debugf("Received external metric query with ns: %s, selector: %s, metricName: %s", namespace, metricSelector.String(), info.Metric)
```

You'll see that `info.Metric` is already lowercase, this is inherent to how the `ExternalMetrics` is implemented in Kubernetes, and impact any `ExternalMetric` provider, not only Datadog Cluster Agent, see https://github.com/kubernetes/kubernetes/issues/72996 for more explanation.

sebbrandt87 replied:
Hey @vboulineau,
okay we might come to the real root cause 😄 . Could it be that [this](https://github.com/DataDog/datadog-agent/pull/3013) logic got removed in the meantime and so the implementation is broken because of that? It seems that I haven't seen recognized https://github.com/kubernetes/kubernetes/issues/72996. 👍

sebbrandt87 replied:
If I request with the lowercased metric name `domain.concurrentrequests`

```bash
kubectl get --raw "/apis/external.metrics.k8s.io/v1beta1/namespaces/search/domain.concurrentrequests?labelSelector=env%3Dlive" | jq .
```

 I get this result

```json
{
      "metricName": "domain.concurrentRequests",
      "metricLabels": {
        "env": "live"
      },
      "timestamp": "2022-01-26T11:58:47Z",
      "value": "0"
    }
```

so it is 100% true, also related to kubernetes/kubernetes#72996 that we only can query the kubernetes extern.metrics.k8s.io API in this way.

sebbrandt87 replied:
so what I did now, because I cannot get local unit tests to work at the moment, changed `queryMetricName: "domain.concurrentrequests",` to be lowercase.

vboulineau replied:
If you don't see the `DEBUG` logs above while `DEBUG` logs are activated, it means you are using the legacy implementation (without `DatadogMetrics` CRD at all), so you're no going through the code path you updated here, but to the legacy one: `pkg/clusteragent/custommetrics`, in which case it should works, so I'm not clear on the configuration you use.

There is a way we could support this in autogen `DatadogMetrics` but it's not straightforward and not easy to understand the limitations:
- If you have two metrics that are only different due to case (e.g. `FOO.bar` and `foo.bar`), one of the two will "win" (we cannot predict which one), and the "winner" value will silently be used for both.
- If you create an HPA with a wrong metric name (`Foo.bar` instead of `foo.bar` for instance), we'll need to change the Autogen metrics query while keeping the same id, which was not intended in the original design.
- If you have two HPAs referencing the same metrics+labels but you made a mistake in one (like `Foo.bar` and `foo.bar`), it will work 50% of the time (same value for both) and fail 50% of the time (no value for both).

While not common, these issues are surprising and difficult to debug.

These issues are present in the `legacy` implementation and it's impossible to fix them as we always get a lower case `metricName` from Kubernetes.
For these reasons I strongly recommend to go the safest way, which is to use `DatadogMetrics` option, which does not have any of these issues.

sebbrandt87 replied:
I have understood the issue already @vboulineau 🙇 
I just try to implement, out of a customer perspective, the DataDog [documented](https://docs.datadoghq.com/metrics/custom_metrics/#naming-custom-metrics) behaviour and rule.

So the way to go is in my opinion: 
> If you create an HPA with a wrong metric name (Foo.bar instead of foo.bar for instance), we'll need to change the Autogen metrics query while keeping the same id, which was not intended in the original design.

I am happy to help with that. I would for now set this PR as WIP again and see in the next days how I find the time to proceed, if this is okay for you?

vboulineau replied:
I'll circle back internally on what's the best solution, but I'm not sure we want to change the current behaviour.

sebbrandt87 replied:
any news here @vboulineau ?

vboulineau replied:
Hey @sebbrandt87,

We chose to keep the current implementation. We'll document the limitation on Autogen metrics and recommend to use `DatadogMetric` directly when using case sensitive metrics name.

AliDatadog commented:
Hi @sebbrandt87,
Thank you for this contribution. Closed as we decided to keep the current implementation.
